src/libsystemd-network/network-internal.c

```diff
@@ -396,7 +396,7 @@ int config_parse_iaid(const char *unit,
                 return 0;
         }
 
-        *((uint32_t *)data) = iaid;
+        *((int64_t *)data) = iaid;
```

Comment:
I liked the approach of a separate `bool` to know if the value was enabled or not that was used in 0e408b82b8bd much better. Especially the approach of using two different mechanisms in two parts of the code feel wrong. I think this patch should be changed to take `Network` as `data`, and a `bool` should be added in `Network` to signal whether the iaid was set. @poettering, @yuwata opinions?

Reply:
(In particular, casting a signed integer to an integer of smaller width is implementation-defined if the original value cannot be represented exactly by the narrower type. This patch does not invoke this implementation-defined behaviour, but nonetheless, I would prefer a more straightforward implementation with a separate variable.)

Reply:
I didn't do that in this patch (at two places) because:

- in `config_parse_iaid()`, I would need to introduce a struct, like `struct { uint32_t val; bool has; }` and pass that on. I didn't see precedence for such a thing so I hijacked a int64_t.

- in `sd_dhcp_client_set_iaid_duid*()`, these functions already have so many arguments, so adding yet another `bool use_iaid` seem less preferable. Also, as the original value from the config is already stored as `int64_t`, letting the caller unpack that in two parameters was not great.

On the other hand, in 0e408b82b8bd7675234cf58009475d4f4c0a491a the bool field fits in an existing gap of the structure (alignment wise), so it had no extra overhead -- compared to a aligned int64_t field.

But I don't mind either way.

Reply:
> (In particular, casting a signed integer to an integer of smaller width is implementation-defined if the original value cannot be represented exactly by the narrower type. 

It casts an unsigned fixed width integer (`uint32_t`) to a signed fixed width integer of smaller range (`int64_t`). I think, according to C99 (which introduces these typedefs) this is always well defined.

> I would prefer a more straightforward implementation with a separate variable.)

OK. Does this mean, to typedef a struct? Should the struct be generic (like `OptionalU32` or `NullableU32`) or specifically named `IAID`?

Reply:
See `config_parse_dhcp_route_table`.  You might need to move `config_parse_iaid` to `networkd-network.c` too. The function itself should be changed along those lines:
```diff
diff --git src/libsystemd-network/network-internal.c src/libsystemd-network/network-internal.c
index e889fb7f62..329e34753a 100644
--- src/libsystemd-network/network-internal.c
+++ src/libsystemd-network/network-internal.c
@@ -381,13 +381,14 @@ int config_parse_iaid(const char *unit,
                       const char *rvalue,
                       void *data,
                       void *userdata) {
+        Network *network = data;
         uint32_t iaid;
         int r;
 
         assert(filename);
         assert(lvalue);
         assert(rvalue);
-        assert(data);
+        assert(network);
 
         r = safe_atou32(rvalue, &iaid);
         if (r < 0) {
@@ -396,7 +397,8 @@ int config_parse_iaid(const char *unit,
                 return 0;
         }
 
-        *((int64_t *)data) = iaid;
+        network->iaid = iaid;
+        network->iaid_set = true;
 
         return 0;
 }
diff --git src/network/networkd-network-gperf.gperf src/network/networkd-network-gperf.gperf
index 8bbdd290ad..e58ec4f0a9 100644
--- src/network/networkd-network-gperf.gperf
+++ src/network/networkd-network-gperf.gperf
@@ -135,7 +135,7 @@ DHCP.DUIDRawData,                       config_parse_duid_rawdata,
 DHCP.RouteMetric,                       config_parse_unsigned,                          0,                             offsetof(Network, dhcp_route_metric)
 DHCP.RouteTable,                        config_parse_dhcp_route_table,                  0,                             0
 DHCP.UseTimezone,                       config_parse_bool,                              0,                             offsetof(Network, dhcp_use_timezone)
-DHCP.IAID,                              config_parse_iaid,                              0,                             offsetof(Network, iaid)
+DHCP.IAID,                              config_parse_iaid,                              0,                             0
 DHCP.ListenPort,                        config_parse_uint16,                            0,                             offsetof(Network, dhcp_client_port)
 DHCP.RapidCommit,                       config_parse_bool,                              0,                             offsetof(Network, rapid_commit)
 DHCP.ForceDHCPv6PDOtherInformation,     config_parse_bool,                              0,                             offsetof(Network, dhcp6_force_pd_other_information)
```